### PR TITLE
Add name_func optional attribute for asyncpg adapter

### PR DESCRIPTION
**Pull request by teror4uks**
I faced an issue related to pg bouncer and prepared statement cache flow in asyncpg dialect. Regarding this discussion https://github.com/sqlalchemy/sqlalchemy/issues/6467 I prepared PR to support an optional parameter `name` in prepared statement which is allowed, since 0.25.0 version in `asyncpg` https://github.com/MagicStack/asyncpg/pull/846

**UPD:**
the issue with proposal: https://github.com/sqlalchemy/sqlalchemy/issues/9608

### Description
Added optional parameter `name_func` to `AsyncAdapt_asyncpg_connection` class which will call on the `self._connection.prepare()` function and populate a unique name.

so in general instead this

```python

from uuid import uuid4

from asyncpg import Connection


class CConnection(Connection):
    def _get_unique_id(self, prefix: str) -> str:
        return f'__asyncpg_{prefix}_{uuid4()}__'

engine = create_async_engine(...,     
    connect_args={
        'connection_class': CConnection,
    },
)

```

would be enough 

```python
from uuid import uuid4

engine = create_async_engine(...,     
    connect_args={
        'name_func': lambda:  f'__asyncpg_{uuid4()}__',
    },
)

```

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by CaselIT** [comment]
Hi, can you please also open an issue to properly track this?

File: lib/sqlalchemy/dialects/postgresql/asyncpg.py [hunk 1]
```diff
@@ -644,11 +644,12 @@ class AsyncAdapt_asyncpg_connection(AdaptedConnection):
         "_prepared_statement_cache",
         "_invalidate_schema_cache_asof",
         "_execute_mutex",
+        "_name_func"
     )
 
     await_ = staticmethod(await_only)
 
-    def __init__(self, dbapi, connection, prepared_statement_cache_size=100):
+    def __init__(self, dbapi, connection, prepared_statement_cache_size=100, name_func=None):
```
Comment (CaselIT) [inline lib/sqlalchemy/dialects/postgresql/asyncpg.py:652]: I think name_func is too generic, expecially on the engine.

Let's call this `pepared_statement_name_func` or similar (@zzzeek suggestions here?)

**Comment by teror4uks** [comment]
Added the issue with proposal https://github.com/sqlalchemy/sqlalchemy/issues/9608, updated the description.

File: lib/sqlalchemy/dialects/postgresql/asyncpg.py (hunk 1 above)
Comment (teror4uks) [reply lib/sqlalchemy/dialects/postgresql/asyncpg.py:652]: yeah for sure, my naming choice is quite wide

Comment (zzzeek) [reply lib/sqlalchemy/dialects/postgresql/asyncpg.py:652]: I would use prepared_statement_name_func.

here is the problem.   this scheme will create endless numbers of prepared statements on pooled connections that are not reused.   I think if one is using this scheme, it's very important that [discard](https://www.postgresql.org/docs/current/sql-discard.html) is used appropriately to at the very least release prepared statements ( as well as transactional state as always)   I guess if pgbouncer is used, then on the client (here) we are using NullPool and pgbouncer will have DISCARD set up appropriately but we would to make sure this is mentioned (1. use nullpool w/ pgbouncer; 2. make sure DISCARD is set up).

@teror4uks is that the configuration you are using?    basically with this param I want to have an understanding of a working context for it.

Comment (teror4uks) [reply lib/sqlalchemy/dialects/postgresql/asyncpg.py:652]: yes you are right we are using NullPool and pgbouncer, the common setup looks like this 

```python
engine = create_async_engine(
    url,
    poolclass=NullPool,
    connect_args={
        'statement_cache_size': 0,
        'connection_class': CConnection,
    },
)
```

Comment (zzzeek) [reply lib/sqlalchemy/dialects/postgresql/asyncpg.py]: OK so when we document this, we should put that example in place and definitely mention that these prepared statements will build up like lightning if they are not cleared out

**Review by CaselIT — APPROVED** [review]
Thanks

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision bd130febabbfae4738dc227db7e97cb0921f4488 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change bd130febabbfae4738dc227db7e97cb0921f4488: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Comment by teror4uks** [comment]
I've seen few pep issues related to lines longer than 79 or 80 letters so fixed it in the latest commit.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e1e5b8d66caab8ab28b9cf053fe39588510d1ac4 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset e1e5b8d66caab8ab28b9cf053fe39588510d1ac4 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Comment by teror4uks** [comment]
it seems at the first time I was too optimistic that fixed all issues, flake and black linters found something else, so I ran the same command locally before committing changes again seems now all fixed :see_no_evil: 

```bash
❯ tox -r -e pep8
......
pep8: commands[7]> python ./tools/sync_test_files.py --check
pep8: commands[8]> python ./tools/generate_sql_functions.py --check
.pkg: _exit> python /home/t4ks/projects/sqlalchemy/venv/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  pep8: OK (155.95=setup[52.72]+cmd[75.62,0.50,0.93,0.87,5.90,11.67,5.07,0.40,2.26] seconds)
  congratulations :) (156.11 seconds)
```

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision f2bdac62b20dc6d3afa34e56fa28ae5290a92833 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset f2bdac62b20dc6d3afa34e56fa28ae5290a92833 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Comment by CaselIT** [comment]
thanks for the update!

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

OK code is fine, now the hard part :)   documentation, quick example of setup, changelog file.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Comment by CaselIT** [comment]
@teror4uks Let us know if you want to give it a go at writing some docs and the changelog, otherwise we can take it over

**Comment by teror4uks** [comment]
@CaselIT  I wouldn't mind to add it, here is what I wrote (not yet committed, maybe you have some suggestions)

I assume it should be [here](https://github.com/teror4uks/sqlalchemy/blob/f2bdac62b20dc6d3afa34e56fa28ae5290a92833/lib/sqlalchemy/dialects/postgresql/asyncpg.py#L119-L119)

```
Prepared Statement Name
-----------------------

By default, asyncpg enumerates prepared statements in numeric order, which 
can lead to errors if a name has already been taken for another prepared 
statement. This issue can arise if your application uses database proxies 
such as PgBouncer to handle connections. One possible workaround is to 
use dynamic prepared statement names, which asyncpg now supports through 
an optional name value for the statement name. This allows you to 
generate your own unique names that won't conflict with existing ones. 
To achieve this, you can provide a function that will be called every time 
a prepared statement is prepared::

    from uuid import uuid4

    engine = create_async_engine(
        "postgresql+asyncpg://user:pass@hostname/dbname",     
        poolclass=NullPool,
        connect_args={
            'pepared_statement_name_func': lambda:  f'__asyncpg_{uuid4()}__',
        },
    )

.. seealso::

   https://github.com/MagicStack/asyncpg/issues/837

   https://github.com/sqlalchemy/sqlalchemy/issues/6467

.. warning:: To prevent a buildup of useless prepared statements in 
   your application, it's important to use the ``NullPool`` poolclass and 
   PgBouncer with a configured `DISCARD https://www.postgresql.org/docs/current/sql-discard.html`_ 
   setup. The DISCARD command is used to release resources held by the db connection, 
   including prepared statements. Without proper setup, prepared statements can 
   accumulate quickly and cause performance issues.


```

**Comment by CaselIT** [comment]
looks ok to me, thanks.

I would put in in the paragraph above that's also talking about the statement cache, so here https://github.com/teror4uks/sqlalchemy/blob/f2bdac62b20dc6d3afa34e56fa28ae5290a92833/lib/sqlalchemy/dialects/postgresql/asyncpg.py#L100

**Comment by teror4uks** [comment]
@CaselIT  done committed changes

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision b4bc8d3e57ab095a26112830ad4bea36083454e3 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset b4bc8d3e57ab095a26112830ad4bea36083454e3 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

File: lib/sqlalchemy/dialects/postgresql/asyncpg.py [hunk 2]
```diff
@@ -98,6 +98,44 @@
    stale, nor can it retry the statement as the PostgreSQL transaction is
    invalidated when these errors occur.
 
+.. _asyncpg_prepared_statement_name:
+
+Prepared Statement Name
+-----------------------
+
+By default, asyncpg enumerates prepared statements in numeric order, which
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/postgresql/asyncpg.py:106]: **mike bayer** (zzzeek) wrote:

this is great!  nice job, well written

want some more doc tasks ?  :)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

two quickies in the changelog , besides that it's ready to go

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

* doc/build/changelog/unreleased_20/9608.rst (line 5): typo
* doc/build/changelog/unreleased_20/9608.rst (line 8): we could use a seealso here since there's no paramref we can link to

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

changelog quickies

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Comment by CaselIT** [comment]
I'll fix later today

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

alrighty

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4565 has been **merged**. Congratulations! :)

File: lib/sqlalchemy/dialects/postgresql/asyncpg.py (hunk 2 above)
Comment (teror4uks) [reply lib/sqlalchemy/dialects/postgresql/asyncpg.py:106]: hah thanks, well I might help if find the time :D

**Pull request closed without merging**